components/dashboard/src/components/running-prebuild-view.tsx

```diff
@@ -36,28 +36,40 @@ export class RunningPrebuildView extends React.Component<RunningPrebuildViewProp
         this.state = {};
         this.logline = `Workspace prebuild is ${this.props.justStarting} ...\r\n`;
 
-        let buildIsDone = false;
         this.props.service.registerClient({
             onHeadlessWorkspaceLogs: evt => {
                 if (evt.workspaceID !== this.props.prebuildingWorkspaceId) {
                     return;
                 }
 
+                // TODO(ak) there is no contract that setState should immediately trigger render
+                // it maybe buffer state and then call render, so some log lines can be lost
+                // we should rather get a reference to xterm in this component and write here directly
                 this.logline = evt.text;
                 this.setState({ updateLoglineToggle: !this.state.updateLoglineToggle });
-
-                if (!buildIsDone && !HeadlessWorkspaceEventType.isRunning(evt.type)) {
-                    buildIsDone = true;
-                    this.props.onBuildDone(HeadlessWorkspaceEventType.didFinish(evt.type));
-                }
-            },
-            onInstanceUpdate: () => {},
-            onWorkspaceImageBuildLogs: () => {}
+            }
         });
     }
 
+    private readonly toDispose = new DisposableCollection();
     componentWillMount() {
-        this.props.service.server.watchHeadlessWorkspaceLogs(this.props.prebuildingWorkspaceId);
+        this.watch();
+        this.toDispose.push(this.props.service.registerClient({
+            notifyDidOpenConnection: () => this.watch()
+        }));
+    }
+    componentWillUnmount() {
+        this.toDispose.dispose();
+    }
+
+    private watch(): void {
+        this.props.service.server.watchHeadlessWorkspaceLogs(this.props.prebuildingWorkspaceId).catch(e => {
+            if (e instanceof ConnectionError) {
+                return;
```

Comment:
what does `ConnectionError` mean in that case?   `onWatchPrebuild ` will be triggered anyways, thus `doCreateWorkspaceWithMode(CreateWorkspaceMode.UsePrebuild)` is called as well.

Reply:
though, it's likely that I just need more ☕

Reply:
all these `instanceof ConnectionError` don't always work and I'm going to remove them